crates/wasm-smith/src/lib.rs

```diff
@@ -1042,6 +1057,161 @@ where
+            }
+            let mut exports = Vec::with_capacity(module.exports.len());
+            for (name, item) in module.exports.iter() {
+                let ty = module.type_of(item);
+                exports.push((name.clone(), ty));
+            }
+            let ty = Rc::new(ModuleType {
+                imports,
+                exports: Rc::new(InstanceType { exports }),
+            });
+
+            // And then given the type of the module we copy it over to
+            // ourselves and get a type index. This index goes into the module
+            // section and we also record the type of this module in our
+            // `modules` array.
+            let ty_idx = self.get_module_type_idx(&ty);
+            modules.push(ty_idx);
+            self.modules.push(self.module_type(ty_idx).clone());
+            self.defined_modules.push(module);
+            Ok(true)
+        })?;
+        if !modules.is_empty() || u.arbitrary()? {
+            self.initial_sections.push(InitialSection::Module(modules));
+        }
+        Ok(())
+    }
+
+    /// Copies the type `ty`, possibly from a different module, into this
+    /// module.
+    fn copy_type_to_self(&mut self, ty: &EntityType) -> EntityType {
+        match ty {
+            // Globals, tables, and memories are easy, it's just copying some
+            // bits around.
+            EntityType::Global(t) => EntityType::Global(t.clone()),
+            EntityType::Table(t) => EntityType::Table(t.clone()),
+            EntityType::Memory(t) => EntityType::Memory(t.clone()),
+
+            // Functions, instances, and modules are hairier, however, since
+            // they chases pointers. We delegate to some helper methods for
+            // each one.
+            EntityType::Func(_other_idx, sig) => {
+                EntityType::Func(self.get_func_type_idx(sig), sig.clone())
+            }
+            EntityType::Instance(_other_idx, sig) => {
+                EntityType::Instance(self.get_instance_type_idx(sig), sig.clone())
+            }
+            EntityType::Module(_other_idx, sig) => {
+                EntityType::Module(self.get_module_type_idx(sig), sig.clone())
+            }
+        }
+    }
+
+    fn get_func_type_idx(&mut self, sig: &Rc<FuncType>) -> u32 {
+        // First check to see if our module already defines this function
+        // signature...
+        if let Some((i, _)) = self.func_types().find(|(_, ty)| **ty == **sig) {
+            return i;
+        }
+        // ... and otherwises add a new function signature.
+        let type_idx = self.types.len() as u32;
+        self.func_types.push(type_idx);
+        self.push_type(Type::Func(sig.clone()));
+        type_idx
+    }
+
+    fn get_instance_type_idx(&mut self, sig: &Rc<InstanceType>) -> u32 {
+        // For simplicity for now we always inject a new `instance` type in
+        // this situation instead of looking for one that's already defined.
```

Comment:
Hash consing shouldn't make this too hard either.

Reply:
I can implement this but it's kind of annoying to do so since it's a lot of derives and it doesn't fit super well with how the types are currently represented which also have indices (which aren't hashed). I'll skip this for now since it seems likely that the module linking proposal will soon be modified that the type of a module isn't declared, the module is simply defined, which would make this an unnecessary operation anyway.